### PR TITLE
Register when an exception was handled or not

### PR DESCRIPTION
**Pull request by ibm5155**
With this Pr, Sentry.io is going to register if an exception was handled by setting the handled flag in the Mechanism class.

It will not fix the "event handled" internal flag because I'm unsure which parameters I need to inform for that to work.

### CODE REVIEW

File: test/Sentry.Tests/AppDomainUnhandledExceptionIntegrationTests.cs [hunk 1]
```diff
@@ -56,6 +59,27 @@ public void Handle_TerminatingTrue_DisposesHub()
...
+            Assert.True(exception.Data?.Contains("Sentry:Mechanism"));
+
+
+            var stackTraceFactory = Substitute.For<ISentryStackTraceFactory>();
+            var exceptionProcessor = new MainExceptionProcessor(SentryOptions, () => stackTraceFactory);
+            var @event = new SentryEvent(exception);
+
+            exceptionProcessor.Process(exception, @event);
+            Assert.NotNull(@event.SentryExceptions.ToList().FirstOrDefault(p => p.Mechanism?.Handled == false));
+
```
Comment (bruno-garcia) [inline test/Sentry.Tests/AppDomainUnhandledExceptionIntegrationTests.cs]:
```suggestion
```

File: src/Sentry/Internal/MainExceptionProcessor.cs [hunk 2]
```diff
@@ -110,14 +110,24 @@ internal static Mechanism GetMechanism(Exception exception)
         {
             Debug.Assert(exception != null);
 
-            Mechanism mechanism = null;
+            Mechanism mechanism = new Mechanism()
```
Comment (bruno-garcia) [inline src/Sentry/Internal/MainExceptionProcessor.cs]:
```suggestion
            var mechanism = new Mechanism()
```

File: src/Sentry/Internal/MainExceptionProcessor.cs [hunk 3]
```diff
@@ -110,14 +110,24 @@ internal static Mechanism GetMechanism(Exception exception)
...
+            {
+                Handled = true
```
Comment (bruno-garcia) [inline src/Sentry/Internal/MainExceptionProcessor.cs]: Can we leave this `null` by default instead of defaulting to `true`?
Not sure if Sentry expects it to be explicitly set to `true` though

File: src/Sentry/Internal/MainExceptionProcessor.cs [hunk 4]
```diff
@@ -110,14 +110,24 @@ internal static Mechanism GetMechanism(Exception exception)
...
+            }
+            if (exception.Data?.Contains("Sentry:Handled") == true)
+            {
+                mechanism.Handled = (exception.Data["Sentry:Handled"] as bool?).Value;
+                exception.Data.Remove("Sentry:Handled");
+            }
+            if (exception.Data?.Contains("Sentry:Mechanism") == true)
+            {
+                mechanism.Type = exception.Data["Sentry:Mechanism"] as string;
+                exception.Data.Remove("Sentry:Mechanism");
```
Comment (bruno-garcia) [inline src/Sentry/Internal/MainExceptionProcessor.cs]: Now we have two lookups in the hashtable:

```
            if (exception.Data?.Contains("Sentry:Handled") == true)
```

And then a second one:

```
                 mechanism.Handled = (exception.Data["Sentry:Handled"] as bool?).Value;
```

Ideally we'd use `TryGetValue` but since this is the old school `IDictionary`, it's not available.
We could at least use the indexers and pattern matching:

```suggestion
        // Throw these at the top of the Mechanism class so it can be used by AppDomainUnhandledExceptionIntegration.cs too
        // Can be used from the unit tests too
        public static readonly string HandledKey = "Sentry:Handled";
        public static readonly string MechanismKey = "Sentry:Mechanism";
        
            if (exception.Data[HandledKey] is bool handled)
            {
                mechanism.Handled = handled;
                exception.Data.Remove(HandledKey);
            }
            if (exception.Data[MechanismKey] is string mechanismName)
            {
                mechanism.Type = mechanismName;
                exception.Data.Remove(MechanismKey);
```

File: test/Sentry.Tests/AppDomainUnhandledExceptionIntegrationTests.cs [hunk 5]
```diff
@@ -56,6 +59,27 @@ public void Handle_TerminatingTrue_DisposesHub()
...
 
+        [Fact]
+        public void Handle_TerminatingTrue_IsHandled()
+        {
+            var sut = _fixture.GetSut();
+            sut.Register(_fixture.Hub, SentryOptions);
+
+            var exception = new Exception();
+            sut.Handle(this, new UnhandledExceptionEventArgs(exception, true));
+            Assert.Equal(exception.Data["Sentry:Handled"] as bool?, (bool?)false);
```
Comment (bruno-garcia) [inline test/Sentry.Tests/AppDomainUnhandledExceptionIntegrationTests.cs]: The order here's inverted. Expected is on the left. But could just be:
```suggestion
            Assert.False((bool)exception.Data["Sentry:Handled"]);
```

Also now could use the `static readonly reference` from `Mechanism` instead of the string here

File: test/Sentry.Tests/AppDomainUnhandledExceptionIntegrationTests.cs [hunk 6]
```diff
@@ -56,6 +59,27 @@ public void Handle_TerminatingTrue_DisposesHub()
...
+            Assert.True(exception.Data?.Contains("Sentry:Mechanism"));
```
Comment (bruno-garcia) [inline test/Sentry.Tests/AppDomainUnhandledExceptionIntegrationTests.cs]:
```suggestion
            Assert.True(exception.Data.Contains("Sentry:Mechanism"));
```

File: test/Sentry.Tests/AppDomainUnhandledExceptionIntegrationTests.cs [hunk 7]
```diff
@@ -56,6 +59,27 @@ public void Handle_TerminatingTrue_DisposesHub()
...
+
```
Comment (bruno-garcia) [inline test/Sentry.Tests/AppDomainUnhandledExceptionIntegrationTests.cs]:
```suggestion
```

File: test/Sentry.Tests/AppDomainUnhandledExceptionIntegrationTests.cs [hunk 8]
```diff
@@ -56,6 +59,27 @@ public void Handle_TerminatingTrue_DisposesHub()
...
+
+            var stackTraceFactory = Substitute.For<ISentryStackTraceFactory>();
+            var exceptionProcessor = new MainExceptionProcessor(SentryOptions, () => stackTraceFactory);
+            var @event = new SentryEvent(exception);
+
+            exceptionProcessor.Process(exception, @event);
+            Assert.NotNull(@event.SentryExceptions.ToList().FirstOrDefault(p => p.Mechanism?.Handled == false));
```
Comment (bruno-garcia) [inline test/Sentry.Tests/AppDomainUnhandledExceptionIntegrationTests.cs]:
```suggestion
            Assert.NotNull(@event.SentryExceptions.Single(p => p.Mechanism.Handled == false));
```

File: test/Sentry.Tests/AppDomainUnhandledExceptionIntegrationTests.cs [hunk 9]
```diff
@@ -56,6 +59,27 @@ public void Handle_TerminatingTrue_DisposesHub()
             disposableHub.Received(1).Dispose();
         }
 
+        [Fact]
+        public void Handle_TerminatingTrue_IsHandled()
```
Comment (bruno-garcia) [inline test/Sentry.Tests/AppDomainUnhandledExceptionIntegrationTests.cs]: This tests assert the value Handled is false when app is terminating. Would be nice to have another test that tests it's set to `null` or `true` when this is not captured through the `UnhandledExceptionhandler`.

```suggestion
        public void Handle_TerminatingTrue_IsHandledFalse()
```

File: src/Sentry/Internal/MainExceptionProcessor.cs (hunk 3 above)
Comment (ibm5155) [reply src/Sentry/Internal/MainExceptionProcessor.cs]: image
The handled inside of Mechanism doesn't seems to affect that "handled vs unhandled", it only point that the exception was handled or not.
That's why I was thinking that by default, an exception is handled, unless it comes from the Unhandled Exception Integration.

File: test/Sentry.Tests/Internals/MainExceptionProcessorTests.cs [hunk 10]
```diff
@@ -125,6 +126,33 @@ public void Process_TwoExceptionsWithData_DataOnEventExtra()
...
+        [Fact]
+        public void Process_ExceptionWithout_Handled()
+        {
+            var sut = _fixture.GetSut();
+            var evt = new SentryEvent();
+            var exp = new Exception();
+
+            sut.Process(exp, evt);
+
+            Assert.NotNull(evt.SentryExceptions.ToList().Single(p => p.Mechanism.Handled == null));
```
[the remark this answers is not in the document]
Comment (ibm5155) [reply test/Sentry.Tests/Internals/MainExceptionProcessorTests.cs]: Assert.Single(evt.SentryExceptions, p => p.Mechanism.Handled == null);
doesn't work in this case, the predicate needs to be a SentryException

[the remark this answers is not in the document]
Comment (ibm5155) [reply test/Sentry.Tests/Internals/MainExceptionProcessorTests.cs]: what about 
`Assert.Single(evt.SentryExceptions.Where(p => p.Mechanism.Handled == null));`
?

[the remark this answers is not in the document]
Comment (bruno-garcia) [reply test/Sentry.Tests/Internals/MainExceptionProcessorTests.cs]: 👍

**Comment by bruno-garcia** [comment]
Thanks @ibm5155 !
